udata/core/dataservices/api.py

```diff
@@ -95,6 +97,82 @@ def delete(self, dataservice):
         return "", 204
 
 
+dataservice_add_datasets_fields = api.model(
+    "DataserviceDatasetsAdd",
+    {
+        "id": fields.String(description="Id of the dataset to add", required=True),
+    },
+    location="json",
+)
+
+
+@ns.route("/<dataservice:dataservice>/datasets/", endpoint="dataservice_datasets", doc=common_doc)
+class DataserviceDatasetsAPI(API):
+    @api.secure
+    @api.doc("dataservice_datasets_create")
+    @api.expect([dataservice_add_datasets_fields])
+    @api.marshal_with(Dataservice.__read_fields__)
+    @api.response(400, "Malformed object id(s) in request")
+    @api.response(400, "Expecting a list")
+    @api.response(400, "Expecting a list of dicts with id attribute")
+    @api.response(404, "Dataservice not found")
+    @api.response(403, "Forbidden")
+    @api.response(410, "Dataservice has been deleted")
+    def post(self, dataservice):
+        if dataservice.deleted_at and not (
+            # Allow requests containing "deleted_at: None" to undelete.
+            "deleted_at" in request.json and request.json.get("deleted_at") is None
+        ):
+            api.abort(410, "Dataservice has been deleted")
+
+        OwnablePermission(dataservice).test()
+
+        data = request.json
+
+        if not isinstance(data, list):
+            api.abort(400, "Expecting a list")
+        if not all(isinstance(d, dict) and d.get("id") for d in data):
+            api.abort(400, "Expecting a list of dicts with id attribute")
+
+        try:
+            datasets = Dataset.objects.filter(id__in=[d["id"] for d in data]).only("id")
+            diff = set(d.id for d in datasets) - set(d.id for d in dataservice.datasets)
+        except mongoengine.errors.ValidationError:
+            api.abort(400, "Malformed object id(s) in request")
+
+        if diff:
+            dataservice.datasets += [ObjectId(did) for did in diff]
+            dataservice.save()
+
+        return dataservice, 201
+
+
+@ns.route(
+    "/<dataservice:dataservice>/datasets/<dataset:dataset>/",
+    endpoint="dataservice_dataset",
+    doc=common_doc,
+)
+class DataserviceDatasetAPI(API):
+    @api.secure
+    @api.response(404, "Dataservice not found")
+    @api.response(404, "Dataset not found in dataservice")
+    def delete(self, dataservice, dataset):
+        if dataservice.deleted_at and not (
+            # Allow requests containing "deleted_at: None" to undelete.
+            "deleted_at" in request.json and request.json.get("deleted_at") is None
+        ):
+            api.abort(410, "Dataservice has been deleted")
+
+        OwnablePermission(dataservice).test()
+
+        if dataset.id not in (d.id for d in dataservice.datasets):
+            api.abort(404, "Dataset not found in dataservice")
+        dataservice.datasets = [d for d in dataservice.datasets if d.id != dataset.id]
+        dataservice.save()
+
+        return None, 204
+
+
```

Comment:
I'm a bit confused, is this related to the PR, adding the `by_link` feature?

Reply:
Yes, we add two new endpoint to manipulate the list without sending it (the `by_link` feature allows to not return during reads the list of datasets, but to modify the list we need to have an endpoint to add or remove an item of the list without sending the full list of thousands of datasets)